Add new aws_vpc_endpoint_route_table_association resource

ewbankkit opened the pull request:
This PR follows on from #9244 and addresses #9138.
Here's the log from @kwilczynski's initial commit:
```
This commit adds a new resource which allows to a list of route tables to be
either added and/or removed from an existing VPC Endpoint. This resource would
also be complimentary to the existing `aws_vpc_endpoint` resource where the
route tables might not be specified (not a requirement for a VPC Endpoint to
be created successfully) during creation, especially where the workflow is
such where the route tables are not immediately known.
```
As it stands now the new `aws_vpc_endpoint_route_table_association` resource associates/dissociates as single `route_table_id` with a VPC endpoint.
Somehow I feel it's "better" Terraform to use a `count` attribute on the resource to associate multiple route tables with a VPC endpoint rather than accept a list of route tables in a single association, e.g.
```
resource "aws_vpc" "foo" {
    cidr_block = "10.0.0.0/16"
}

resource "aws_vpc_endpoint" "s3" {
    vpc_id = "${aws_vpc.foo.id}"
    service_name = "com.amazonaws.us-west-2.s3"
}

variable "c" {
    default = 2
}

resource "aws_route_table" "rt" {
    count = "${var.c}"

    vpc_id = "${aws_vpc.foo.id}"

    tags {
        Name = "test"
    }
}

resource "aws_vpc_endpoint_route_table_association" "a" {
    count = "${var.c}"

	vpc_endpoint_id = "${aws_vpc_endpoint.s3.id}"
	route_table_id  = "${element(aws_route_table.rt.*.id, count.index)}"
}
```
Since this is a "synthetic" resource (no underlying AWS resource maps exactly to this) this doesn't cause a cognitive disconnect, plus it greatly simplifies the implementation.
I welcome feedback on this, especially from @coen-hyde and @supergibbs who commented on the associated issue.

ewbankkit commented:
Acceptance tests:
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSVpcEndpointRouteTableAssociation_'
```

jrnt30 commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,189 @@
+	if err != nil {
+		return fmt.Errorf("Error creating VPC Endpoint/Route Table association: %s", err.Error())
+	}
+
```
I believe generally you should be calling the `d.SetId` here to populate the state, the fact that you call `Read` immediately makes it a bit less necessary but is the pattern I've seen around the code base.

ewbankkit replied:
OK, I'll fix that.

ewbankkit commented:
Removing the **[WIP]** marker.
I am considering this as _ready for consideration_.

kwilczynski commented:
@ewbankkit hi there! Thank you for completing this when I could not do it. You are a ⭐️ !

I do agree that relying on `count` seem much more natural with Terraform - my idea of having the ability to pass multiple IDs definitely is not as flexible as using the `count` 👍 .

kwilczynski commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,193 @@
+	return &schema.Resource{
+		Create: resourceAwsVPCEndpointRouteTableAssociationCreate,
+		Read:   resourceAwsVPCEndpointRouteTableAssociationRead,
+		// "resource aws_vpc_endpoint_route_table_association: All fields are ForceNew or Computed w/out Optional, Update is superfluous"
```
I would get rid of the comment here. If we are never going to use `UpdateFunc` then I would not bother with clarification.

ewbankkit commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,193 @@
+			return fmt.Errorf("Error deleting VPC Endpoint/Route Table association: %s", err.Error())
+		}
+
+		switch ec2err.Code() {
```
OK

ewbankkit commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,193 @@
+	return fmt.Sprintf("a-%s%d", endpointId, hashcode.String(rtId))
+}
+
+type vpcEndpointNotFound struct {
```
I'll revisit. I based that on another resource and maybe it's overkill in this situation.

ewbankkit commented:
Pushed changes after @kwilczynski's code review:
* Removed error types and simplified the error handling in `resourceAwsVPCEndpointRouteTableAssociationRead`
* Simplified logging in `resourceAwsVPCEndpointRouteTableAssociationDelete`

stack72 commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,151 @@
+	}
+	if !found {
+		// The association no longer exists.
+		d.SetId("")
```
A quick follow up point for the future - we *usually* log a message here just to let the user know we have removed from state

ewbankkit replied:
OK. I won't change this now. Thanks for the tip.

stack72 commented:
Hi @ewbankkit 

this LGTM! I left a few small questions inline - if you are happy with those then we can push forward

Thanks to @kwilczynski for the review as well

P.

ewbankkit commented:
Sorry, at re:Invent all last week. I'll address these now...

ewbankkit commented on builtin/providers/aws/resource_aws_vpc_endpoint_route_table_association.go:
```diff
@@ -0,0 +1,151 @@
+		return nil, err
+	}
+
+	return output.VpcEndpoints[0], nil
```
Yes. The desired VPC endpoint ID is passed as an input parameter and the [API call](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeVpcEndpoints.html) will return either that VPC endpoint or a [client error](http://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html#api-error-codes-table-client) such as `InvalidVpcEndpointId.Malformed` or `InvalidVpcEndpointId.NotFound`.
We should be good here.

stack72 commented:
Hi @ewbankkit 

Thanks for the work here (and the responses to my nit questions)

This LGTM!

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSVpcEndpointRouteTableAssociation_'                                                                                               ✹
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/05 12:53:35 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSVpcEndpointRouteTableAssociation_ -timeout 120m
=== RUN   TestAccAWSVpcEndpointRouteTableAssociation_basic
--- PASS: TestAccAWSVpcEndpointRouteTableAssociation_basic (50.68s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	50.700s
```
